### ISSUE TITLE
validate scale type

### ISSUE THREAD

**Issue by jfirebaugh** [mapbox/mapbox-gl-style-spec#292]
In v7, style properties have a "function" specification that indicates whether they support either interpolated or piecewise-constant functions. This was never validated, mostly because the two function types have the same syntax (well, nearly -- interpolated functions support `"base"` and piecewise-constant functions don't, which could be validated).

Scale types are a little more involved. We should specify for each style property what scale types are supported:

* ordinal
* power with "normal" rounding (like v7 "interpolated")
* power with "floor" rounding (like v7 "piecewise-constant")

For most style properties this is obvious based on the type, but it's best to be explicit, and I know of at least one outlier: `line-dasharray`, which looks like it might support interpolation but doesn't.

Once this is implemented we should remove the old `"function"` property from the v8 spec.

**Comment by lucaswoj** [comment]
#290 did add scale validation to this repository.

I didn't think it was necessary to specify, per style property, what kind of scales are allowed because we validate the `range` values against the style property. As long as the range values are valid, the scale type isn't important.

For example, we don't need to specify that `line-width` supports `ordinal`, `power, normal`, and `power, floor` because the scale's range is validated such that it only includes numbers. Likewise, for a property like `line-cap`, the range is validated such that it only includes valid enum values.

The only way this could be hairy is if we had enum values that are numbers (which we don't).

**Comment by jfirebaugh** [comment]
Here are examples of invalid scales that should be caught at validation time:

```
"symbol-placement": {
  "domain": [1, 2],
  "range": ["point", "line"],
  "rounding": "normal"
}

"symbol-placement": {
  "domain": [1, 2],
  "range": ["point", "line"],
  "base": 2
}

"line-dasharray": {
  "domain": [1, 2],
  "range": [[0, 1, 2], [3, 4, 5]]
}

"line-dasharray": {
  "domain": [1, 2],
  "range": [[0, 1, 2], [3, 4, 5]],
  "rounding": "normal"
}

"line-dasharray": {
  "domain": [1, 2],
  "range": [[0, 1, 2], [3, 4, 5]],
  "base": 2
}
```

Another reason to include the supported types is that when UI for this starts getting built out, it will need to know what types are supported, and we've found it best not to rely on making inferences from other properties for things like this.

**Comment by jfirebaugh** [comment]
Another case is:

```
"line-width": {
  "domain": [1, 2],
  "range": [5, 10],
  "rounding": "floor",
  "base": 2
}
```

But this is probably best validated by separating 'power with "normal" rounding' and 'power with "floor" rounding' into separate scale types.

**Comment by lucaswoj** [comment]
I agree that examples 1 and 2 should fail validation (and will push a commit shortly)
```
"symbol-placement": {
  "domain": [1, 2],
  "range": ["point", "line"],
  "rounding": "normal"
}
"symbol-placement": {
  "domain": [1, 2],
  "range": ["point", "line"],
  "base": 2
}
```
and I agree that's its desirable for the last one to fail validation (and will push a commit shortly)
```
"line-width": {
  "domain": [1, 2],
  "range": [5, 10],
  "rounding": "floor",
  "base": 2
}
```

I'm not sure what's wrong with these, though
```
"line-dasharray": {
  "domain": [1, 2],
  "range": [[0, 1, 2], [3, 4, 5]]
}
"line-dasharray": {
  "domain": [1, 2],
  "range": [[0, 1, 2], [3, 4, 5]],
  "rounding": "normal"
}
"line-dasharray": {
  "domain": [1, 2],
  "range": [[0, 1, 2], [3, 4, 5]],
  "base": 2
}
```
We can interpolate the array values, as long as all the arrays are the same length. Isn't that how functions worked before?

**Comment by jfirebaugh** [comment]
`"line-dasharray"` does not interpolate -- it transitions (with an animation) directly from one range value to another. This is why it's [marked as "piecewise-constant"](https://github.com/mapbox/mapbox-gl-style-spec/blob/9265ddbe7f1d71dabd374349a62af99af7997efc/reference/v7.json#L866). See https://github.com/mapbox/mapbox-gl-style-spec/issues/237 for discussion.

**Comment by lucaswoj** [comment]
Ah! I didn't have that context. Thanks for the link! I'll chew on this now.

**Comment by lucaswoj** [comment]
Closed by #295
